# Workshop 1: Programming basics with Python

## Chapter 1: Python Warm-up

Currently there are two versions of Python: v2.7 and v3.5. This workshop focuses on Python 3.

### Exercise 1-1: The Python Interface
In the Python script below, you can type Python code to solve exercises. Click on Run or type CTRL+ENTER to run the written Python code.

Instructions
1. Experiment in the IPython Shell; type 5 / 8, for example.
2. Add another line of code to the Python script: print(7 + 10).

In [1]:
# Print 5 + 8
print(5 + 8)

# Print 7 + 10
print(7 + 10) 

# Print 5 / 8
print(5 / 8)

13
17
0.625


### Exercise 1-2: When to use Python?
Python is a pretty versatile language. For which applications can you use Python?

- Option 1: You want to do some quick calculations.
- Option 2: For your new business, you want to develop a database-driven website.
- Option 3: Your boss asks you to clean and analyze the results of the latest satisfaction survey.
- Option 4: All of the above.

Answer: Option 4

### Exercise 1-3: Any comments?
You can add comments to your Python scripts. Comments are important to make sure that you and others can understand what your code is about.

To add comments to your Python script, you can use the # tag. These comments are not run as Python code, so they will not influence your result. As an example, take the comment on the right, # Division; it is completely ignored during execution.

Instructions
1. Above the print(7 + 10), add the comment # Addition.

In [2]:
# Division
print(5 / 8)

# Addition
print(7 + 10)

0.625
17


### Exercise 1-4: Python as a calculator
Python is perfectly suited to do basic calculations. Apart from addition, subtraction, multiplication and division, there is also support for more advanced operations such as:
1. Exponentiation ** - This operator raises the number to its left to the power of the number to its right. For example 4 ** 2 will give 16.
2. Modulo % - This operator returns the remainder of the division of the number to the left by the number on its right. For example 18 % 7 equals 4.

Instructions
1. Suppose you have $100, which you can invest with a 10 percent return each year. After one year, it's 100×1.1=110 dollars, and after two years it's 100×1.1×1.1=121.
2. Add code to calculate how much money you end up with after 7 years, and print the result.

In [6]:
# Addition, subtraction
print(5 + 5)
print(5 - 5)

# Multiplication, division, modulo, and exponentiation
print(3 * 5)
print(10 / 2)
print(18 % 7)
print(4 ** 2)

# How much is your $100 worth after 7 years?
print(100*(1.1**7))

10
0
15
5.0
4
16
194.87171000000012


## Chapter 2: Variables and Types

Variable Characteristics
1. Specific, case-sensitive name
2. Call value through variable name

#### Example: Using variables to calculate BMI

In [4]:
height = 1.79
weight = 68.7

bmi = weight/height**2
print(bmi)

21.44127836209856


So far we've only worked with numerical values such as height and weight. In Python, these numbers all have a specific type. You can verify the type of a value with the type function as follows:

In [5]:
type(bmi)

float

As it may be observed, the value for the bmi is a float which is the type of value used by Python to represent real numbers aka a number which can have both an integer part and a fractional part. Python also has a type for integers (int) as follows:

In [9]:
day_of_the_week=5
type(day_of_the_week)

int

However, to do Data Science you need more than integers and floats. Python has a ton of value types. The most common ones are strings (str) and booleans (bool). A string is Python's way to represent text. You can use both double and single quotes to build a string as you can see from the following examples:

In [16]:
x="body mass index"
y='this works too'

print(x)
print(y)

body mass index
this works too


On the other hand the boolean is a type that can either be True or False equal to Yes and No is everyday language. An example of a very useful application to booleans is to perform filtering operations on your data. There is something special about Python data types. Have a look at the following code:

In [17]:
print(2+3)
print('ab'+'cd')

5
abcd


As it may be observed, the + operator behaves differently for different data types. In the exercise that follows you'll create your first variable and experiment with some of Python's data types.

### Exercise 2-1: Variable Assignments
In Python, a variable allows you to refer to a value with a name. To create a variable use =, like this example:

x = 5

You can now use the name of this variable, x, instead of the actual value, 5. Remember, = in Python means assignment, it doesn't test equality!

Instructions
1. Create a variable savings with the value 100.
2. Check out this variable by typing print(savings) in the script.

In [18]:
# Create a variable savings
savings=100

# Print out savings
print(savings)

100


### Exercise 2-2: Calculations with variables
Remember how you calculated the money you ended up with after 7 years of investing $100? You did something like this:

100 * 1.1 ** 7

Instead of calculating with the actual values, you can use variables instead. The savings variable you've created in the previous exercise represents the $100 you started with. It's up to you to create a new variable to represent 1.1 and then redo the calculations!

Instructions
1. Create a variable growth_multiplier, equal to 1.1.
2. Create a variable, result, equal to the amount of money you saved after 7 years.
3. Print out the value of result.

In [19]:
# Create a variable savings
savings = 100

# Create a variable growth_multiplier
growth_multiplier=1.1

# Calculate result
result=savings*(growth_multiplier**7)

# Print out result
print(result)

194.87171000000012


### Exercise 2-3: Other variable types
In the previous exercise, you worked with two Python data types:
1. int, or integer: a number without a fractional part. savings, with the value 100, is an example of an integer.
2. float, or floating point: a number that has both an integer and fractional part, separated by a point. growth_multiplier, with the value 1.1, is an example of a float.

Next to numerical data types, there are two other very common data types:
1. str, or string: a type to represent text. You can use single or double quotes to build a string.
2. bool, or boolean: a type to represent logical values. Can only be True or False (the capitalization is important!).
    
Instructions
1. Create a new string, desc, with the value "compound interest".
2. Create a new boolean, profitable, with the value True.

In [20]:
# Create a variable desc
desc="compound interest"

# Create a variable profitable
profitable=True

### Exercise 2-4: Operations with other types
Different types behave differently in Python. When you sum two strings, for example, you'll get different behavior than when you sum two integers or two booleans. In the following script some variables with different types have already been created. It's up to you to use them.

Instructions
1. Calculate the product of savings and growth_multiplier. Store the result in year1.
2. What do you think the resulting type will be? Find out by printing out the type of year1.
3. Calculate the sum of desc and desc and store the result in a new variable doubledesc.
4. Print out doubledesc. Did you expect this?

In [21]:
savings = 100
growth_multiplier = 1.1
desc = "compound interest"

# Assign product of growth_multiplier and savings to year1
year1=savings*growth_multiplier

# Print the type of year1
print(type(year1))

# Assign sum of desc and desc to doubledesc
doubledesc=desc+desc

# Print out doubledesc
print(doubledesc)

<class 'float'>
compound interestcompound interest


### Exercise 2-5: Type conversion
Using the + operator to paste together two strings can be very useful in building custom messages. Suppose, for example, that you've calculated the return of your investment and want to summarize the results in a string. Assuming the integer savings and float result are defined, you can try something like this:

This will not work, though, as you cannot simply sum strings and integers/floats. To fix the error, you'll need to explicitly convert the types of your variables. More specifically, you'll need str(), to convert a value into a string. str(savings), for example, will convert the integer savings to a string. Similar functions such as int(), float() and bool() will help you convert Python values into any type.

Instructions
1. Click on Run or type CTRL+ENTER to run the code. Try to understand the error message.
2. Fix the code such that the printout runs without errors; use the function str() to convert the variables to strings.
3. Convert the variable pi_string to a float and store this float as a new variable, pi_float.

In [23]:
# Definition of savings and result
savings = 100
result = 100 * 1.10 ** 7

# Fix the printout
print("I started with $" + savings + " and now have $" + result + ". Awesome!")
# Answer: print("I started with $" + str(savings) + " and now have $" + str(result) + ". Awesome!")

# Definition of pi_string
pi_string = "3.1415926"

# Convert pi_string into float: pi_float
pi_float=float(pi_string)
print(pi_float)

I started with $100 and now have $194.87171000000012. Awesome!
3.1415926


### Exercise 2-6: Can Python handle everything?
Now that you know something more about combining different sources of information, have a look at the four Python expressions below. Which one of these will throw an error? You can always copy and paste this code in the IPython Shell to find out!

Possible Answers
Option 1: "I can add integers, like " + str(5) + " to strings."
Option 2: "I said " + ("Hey " * 2) + "Hey!"
Option 3: "The correct answer to this multiple choice exercise is answer number " + 2
Option 4: True + False

Answer: Option 3

## Chapter 3: Python Lists
By now, you've played around with different data types such as:
- float: real numbers
- int: integer
- str: string, text
- bool: true, false

You can save this values in a variable as shown before and each variable then represent a single value. However, as a data scientist you'll often want to work with many data points. For example: If you want to measure the height of everybody in your family and store this information in Python, it would be inconvenient to store a Python variable for each point you collected. Instead you can store all this information in a Python list. You can build such a list with square brackets as follows:

In [31]:
family_heights=[1.73, 1.68, 1.71, 1.89]
print(family_heights[0])
print(family_heights[1])
print(family_heights[2])
print(family_heights[3])

1.73
1.68
1.71
1.89


A list is a way to give a single name to a collection of values. These values or elements can have different value types including lists. Suppose for example that you want to add the names of your parents and sisters to the family_heights list so that you know which heights belong to whom. You can thrown in some strings without issues as follows:

In [5]:
family_heights=['liz', 1.73, 'emma', 1.68, 'mom', 1.71, 'dad', 1.89]
print(family_heights)

['liz', 1.73, 'emma', 1.68, 'mom', 1.71, 'dad', 1.89]


But as previously mentioned, you can also store list inside list. Thus, inside the family_heights list, sublists may be stored for each family member as follows:

In [ ]:
family_heights=[['liz', 1.73], ['emma', 1.68], ['mom', 1.71], ['dad', 1.89]]
print(family_heights)

### Exercise 3-1: Create a list
As opposed to int, bool etc., a list is a compound data type; you can group values together:

After measuring the height of your family, you decide to collect some information on the house you're living in. The areas of the different parts of your house are stored in separate variables for now, as shown in the script.

Instructions
1. Create a list, areas, that contains the area of the hallway (hall), kitchen (kit), living room (liv), bedroom (bed) and bathroom (bath), in this order. Use the predefined variables.
2. Print areas with the print() function.

In [34]:
# area variables (in square meters)
hall = 11.25
kit = 18.0
liv = 20.0
bed = 10.75
bath = 9.50

# Create list areas
house_meas=[hall, kit, liv, bed, bath]

# Print areas
print(house_meas)

[11.25, 18.0, 20.0, 10.75, 9.5]


### Exercise 3-2: Create list with different types
A list can contain any Python type. Although it's not really common, a list can also contain a mix of Python types including strings, floats, booleans, etc. The printout of the previous exercise wasn't really satisfying. It's just a list of numbers representing the areas, but you can't tell which area corresponds to which part of your house. The code on the right is the start of a solution. For some of the areas, the name of the corresponding room is already placed in front. Pay attention here! "bathroom" is a string, while bath is a variable that represents the float 9.50 you specified earlier.

Instructions
1. Finish the line of code that creates the areas list. Build the list so that the list first contains the name of each room as a string and then its area. In other words, add the strings "hallway", "kitchen" and "bedroom" at the appropriate locations.
2. Print areas again; is the printout more informative this time?

In [35]:
# area variables (in square meters)
hall = 11.25
kit = 18.0
liv = 20.0
bed = 10.75
bath = 9.50

# Adapt list areas
areas = ["hallway", hall, "kitchen", kit, "living room", liv, "bedroom", bed, "bathroom", bath]

# Print areas
print(areas)

['hallway', 11.25, 'kitchen', 18.0, 'living room', 20.0, 'bedroom', 10.75, 'bathroom', 9.5]


### Exercise 3-3: Select the valid list
A list can contain any Python type. But a list itself is also a Python type. That means that a list can also contain a list! Python is getting funkier by the minute, but fear not, just remember the list syntax:

- my_list = [el1, el2, el3]

Can you tell which ones of the following lines of Python code are valid ways to build a list?

Possible Answers
- Option 1: A, B and C
- Option 2: B
- Option 3: B and C
- Option 4: C

Answer: Option 1

### Exercise 3-4: List of lists
As a data scientist, you'll often be dealing with a lot of data, and it will make sense to group some of this data. Instead of creating a flat list containing strings and floats, representing the names and areas of the rooms in your house, you can create a list of lists. The script on the right can already give you an idea. Don't get confused here: "hallway" is a string, while hall is a variable that represents the float 11.25 you specified earlier.

Instructions
1. Finish the list of lists so that it also contains the bedroom and bathroom data. Make sure you enter these in order!
2. Print out house; does this way of structuring your data make more sense?
3. Print out the type of house. Are you still dealing with a list?

In [2]:
# area variables (in square meters)
hall = 11.25
kit = 18.0
liv = 20.0
bed = 10.75
bath = 9.50

# house information as list of lists
house = [["hallway", hall],
         ["kitchen", kit],
         ["living room", liv], ["bedroom", bed], ["bathroom", bath]]

# Print out house
print(house)

# Print out the type of house
print(type(house))

[['hallway', 11.25], ['kitchen', 18.0], ['living room', 20.0], ['bedroom', 10.75], ['bathroom', 9.5]]
<class 'list'>


## Chapter 4: Subsetting Lists
After you've created your very own Python list, you need to know how to access information in the list. Python uses something it calls an index to do this. Let's have a look at the family_heights list again.

In [1]:
family_heights=['liz', 1.73, 'emma', 1.68, 'mom', 1.71, 'dad', 1.89]

The first element in the list has index 0, the second has index 1, and so on. The fact that the index starts at zero is called zero indexing. Suppose that you want to select the height of emma which is the fourth element. To select it you use 3 inside square brackets as follows:

In [38]:
print(family_heights[3])

1.68


You can also count backwards using negative indexes. This is useful if you want to get some elements at the end of your list. To get your dad's height for example you'll need the index -1 or 7 as follows:

In [39]:
print(family_heights[-1])
print(family_heights[7])

1.89
1.89


Apart from indexing there is something called slicing which allows you to select multiple elements from a list. Thus, creating a new list. You can do this by specifying a range using a colon as follows:

In [42]:
print(family_heights[2:6])

['emma', 1.68, 'mom', 1.71]


In general, this is the sintax for slicing: [start(inclusive):end(exclusive)]. The index specified before the colon, where the slice starts, is included while the index specified after the colon, where the slice ends, is excluded. You can also leave out the index before or after the colon. If you leave out the index where the slice should begin, you're telling Python to start the slice from index 0. If you leave out the index where the slice should end, you're telling Python to include all elements up to where the list ends. This is shown in the following example:

In [43]:
print(family_heights[:6])
print(family_heights[2:])

['liz', 1.73, 'emma', 1.68, 'mom', 1.71]
['emma', 1.68, 'mom', 1.71, 'dad', 1.89]


### Exercise 4-1: Subset and conquer
Subsetting Python lists is a piece of cake. Take the code sample below, which creates a list x and then selects "b" from it. Remember that this is the second element, so it has index 1. You can also use negative indexing.

- x = ["a", "b", "c", "d"]
- x[1]
- x[-3] # same result!

Remember the areas list from before, containing both strings and floats? Its definition is already in the script. Can you add the correct code to do some Python subsetting?

Instructions
1. Print out the second element from the areas list (it has the value 11.25).
2. Subset and print out the last element of areas, being 9.50. Using a negative index makes sense here!
3. Select the number representing the area of the living room (20.0) and print it out.

In [44]:
# Create the areas list
areas = ["hallway", 11.25, "kitchen", 18.0, "living room", 20.0, "bedroom", 10.75, "bathroom", 9.50]

# Print out second element from areas
print(areas[1])

# Print out last element from areas
print(areas[-1])

# Print out the area of the living room
print(areas[5])

11.25
9.5
20.0


### Exercise 4-2: Subset and calculate
After you've extracted values from a list, you can use them to perform additional calculations. Take this example, where the second and fourth element of a list x are extracted. The strings that result are pasted together using the + operator:

- x = ["a", "b", "c", "d"]
- print(x[1] + x[3])

Instructions
1. Using a combination of list subsetting and variable assignment, create a new variable, eat_sleep_area, that contains the sum of the area of the kitchen and the area of the bedroom.
2. Print the new variable eat_sleep_area.

In [46]:
# Create the areas list
areas = ["hallway", 11.25, "kitchen", 18.0, "living room", 20.0, "bedroom", 10.75, "bathroom", 9.50]

# Sum of kitchen and bedroom area: eat_sleep_area
eat_sleep_area=areas[3]+areas[-3]

# Print the variable eat_sleep_area
print(eat_sleep_area)

28.75


### Exercise 4-3: Slicing and dicing
Selecting single values from a list is just one part of the story. It's also possible to slice your list, which means selecting multiple elements from your list. Use the following syntax:

- my_list[start:end]

The start index will be included, while the end index is not. The code sample below shows an example. A list with "b" and "c", corresponding to indexes 1 and 2, are selected from a list x:

- x = ["a", "b", "c", "d"]
- x[1:3]

The elements with index 1 and 2 are included, while the element with index 3 is not.

Instructions
1. Use slicing to create a list, downstairs, that contains the first 6 elements of areas.
2. Do a similar thing to create a new variable, upstairs, that contains the last 4 elements of areas.
3. Print both downstairs and upstairs using print().

In [48]:
# Create the areas list
areas = ["hallway", 11.25, "kitchen", 18.0, "living room", 20.0, "bedroom", 10.75, "bathroom", 9.50]

# Use slicing to create downstairs
downstairs=areas[0:6]

# Use slicing to create upstairs
upstairs=areas[6:]

# Print out downstairs and upstairs
print(downstairs)
print(upstairs)

['hallway', 11.25, 'kitchen', 18.0, 'living room', 20.0]
['bedroom', 10.75, 'bathroom', 9.5]


### Exercise 4-4: Slicing and dicing (2)
The previous section discussed the syntax where you specify both where to begin and end the slice of your list:

- my_list[begin:end]

However, it's also possible not to specify these indexes. If you don't specify the begin index, Python figures out that you want to start your slice at the beginning of your list. If you don't specify the end index, the slice will go all the way to the last element of your list. To experiment with this, try the following commands in the IPython Shell:

- x = ["a", "b", "c", "d"]
- x[:2]
- x[2:]
- x[:]

Instructions
1. Create downstairs again, as the first 6 elements of areas. This time, simplify the slicing by omitting the begin index.
2. Create upstairs again, as the last 4 elements of areas. This time, simplify the slicing by omitting the end index.

In [ ]:
# Create the areas list
areas = ["hallway", 11.25, "kitchen", 18.0, "living room", 20.0, "bedroom", 10.75, "bathroom", 9.50]

# Alternative slicing to create downstairs
downstairs=areas[:6]

# Alternative slicing to create upstairs
upstairs=areas[6:10]

### Exercise 4-5: Subsetting lists of lists
You saw before that a Python list can contain practically anything; even other lists! To subset lists of lists, you can use the same technique as before: square brackets. Try out the commands in the following code sample in the IPython Shell:

- x = [["a", "b", "c"],
     ["d", "e", "f"],
     ["g", "h", "i"]]
- x[2][0]
- x[2][:2]

x[2] results in a list, that you can subset again by adding additional square brackets. What will house[-1][1] return? house, the list of lists that you created before, is already defined for you in the workspace. You can experiment with it in the IPython Shell.

Possible Answers:
- Option 1: A float: the kitchen area
- Option 2: A string: "kitchen"
- Option 3: A float: the bathroom area
- Option 4: A string: "bathroom"

In [4]:
print(house)
print(house[-1][1])

[['hallway', 11.25], ['kitchen', 18.0], ['living room', 20.0], ['bedroom', 10.75], ['bathroom', 9.5]]
9.5


Answer: Option 3

## Chapter 5: Manipulating Lists
These are ways to change, add, or remove elements in your list

### Changing list elements

In [8]:
fam = ["liz", 1.73, "emma", 1.68, "mom", 1.71, "dad", 1.89]
print(fam)

#Changing dad's height
fam[7]=1.86
print(fam)

#Changing more than one value
fam[0:2] = ["lizzie", 1.74]
print(fam)

['liz', 1.73, 'emma', 1.68, 'mom', 1.71, 'dad', 1.89]
['liz', 1.73, 'emma', 1.68, 'mom', 1.71, 'dad', 1.86]
['lizzie', 1.74, 'emma', 1.68, 'mom', 1.71, 'dad', 1.86]


### Adding and removing elements

In [23]:
#Adding myself to the list
fam_ex = fam + ["me", 1.62]
print(fam_ex)

#Removing myself from the list
del(fam_ex[8:])
print(fam_ex)

['lizzie', 1.74, 'emma', 1.68, 'mom', 1.71, 'dad', 1.86, 'me', 1.62]
['lizzie', 1.74, 'emma', 1.68, 'mom', 1.71, 'dad', 1.86]


### Behind the scenes

What happends when you create a list like: x = ["a", "b", "c"]?
Your storing a list in your computer memory and the address of the list (where the list is in your computer memory in x). This means that x does not actually contain all the list elements. It rather contains a reference to the list. This concept becomes important when you start copying lists.

#### Example:

In [24]:
x = ["a", "b", "c"]
print(x)

y = x
print(y)

y[1] = "z"
print(y)
print(x)

['a', 'b', 'c']
['a', 'b', 'c']
['a', 'z', 'c']
['a', 'z', 'c']


You can see that since x was copied with the equal sign to y, what was actually copied was the reference to the list not the actual values themselves. Thus, when your updating the elements on the list x or y, it's one in the same list in the computer memory your changing. Both x and y point to the same list.

If you want to create a list that points to a new list and contains the same values of x, you'll need to use something other than the equal sign as follows:

In [28]:
x = ["a", "b", "c"]

y = list(x)
y[1] = "z"
print(y)

z = x[:]
print(z)

['a', 'z', 'c']
['a', 'b', 'c']


### Exercise 5-1: Replace list elements
Replacing list elements is pretty easy. Simply subset the list and assign new values to the subset. You can select single elements or you can change entire list slices at once.

Use the IPython Shell to experiment with the commands below. Can you tell what's happening and why?

- x = ["a", "b", "c", "d"]
- x[1] = "r"
- x[2:] = ["s", "t"]

For this and the following exercises, you'll continue working on the areas list that contains the names and areas of different rooms in a house.

Instructions
1. Update the area of the bathroom area to be 10.50 square meters instead of 9.50.
2. Make the areas list more trendy! Change "living room" to "chill zone".

In [30]:
# Create the areas list
areas = ["hallway", 11.25, "kitchen", 18.0, "living room", 20.0, "bedroom", 10.75, "bathroom", 9.50]

# Correct the bathroom area
areas[9]=10.50

# Change "living room" to "chill zone"
areas[4]="chill zone"

print(areas)

['hallway', 11.25, 'kitchen', 18.0, 'chill zone', 20.0, 'bedroom', 10.75, 'bathroom', 10.5]


### Exercise 5-2: Extend a list
If you can change elements in a list, you sure want to be able to add elements to it, right? You can use the + operator:

- x = ["a", "b", "c", "d"]
- y = x + ["e", "f"]

You just won the lottery, awesome! You decide to build a poolhouse and a garage. Can you add the information to the areas list?

Instructions
1. Use the + operator to paste the list ["poolhouse", 24.5] to the end of the areas list. Store the resulting list as areas_1.
2. Further extend areas_1 by adding data on your garage. Add the string "garage" and float 15.45. Name the resulting list areas_2.

In [32]:
# Create the areas list and make some changes
areas = ["hallway", 11.25, "kitchen", 18.0, "chill zone", 20.0,
         "bedroom", 10.75, "bathroom", 10.50]

# Add poolhouse data to areas, new list is areas_1
areas_1 = areas + ["poolhouse", 24.5]

# Add garage data to areas_1, new list is areas_2
areas_2 = areas_1 + ["garage", 15.45]

print(areas_1)
print(areas_2)

['hallway', 11.25, 'kitchen', 18.0, 'chill zone', 20.0, 'bedroom', 10.75, 'bathroom', 10.5, 'poolhouse', 24.5]
['hallway', 11.25, 'kitchen', 18.0, 'chill zone', 20.0, 'bedroom', 10.75, 'bathroom', 10.5, 'poolhouse', 24.5, 'garage', 15.45]


### Exercise 5-3: Delete list elements
Finally, you can also remove elements from your list. You can do this with the del statement:

- x = ["a", "b", "c", "d"]
- del(x[1])

Pay attention here: as soon as you remove an element from a list, the indexes of the elements that come after the deleted element all change!

The updated and extended version of areas that you've built in the previous exercises is coded below. You can copy and paste this into the IPython Shell to play around with the result.

- areas = ["hallway", 11.25, "kitchen", 18.0,
        "chill zone", 20.0, "bedroom", 10.75,
         "bathroom", 10.50, "poolhouse", 24.5,
         "garage", 15.45]
         
There was a mistake! The amount you won with the lottery is not that big after all and it looks like the poolhouse isn't going to happen. You decide to remove the corresponding string and float from the areas list. The ; sign is used to place commands on the same line. The following two code chunks are equivalent:

Which of the code chunks will do the job for us?

Possible Answers:
- Option 1: del(areas[10]); del(areas[11])
- Option 2: del(areas[10:11])
- Option 3: del(areas[-4:-2])
- Option 4: del(areas[-3]); del(areas[-4])

Answer: Option 3

In [33]:
print(areas_2)

del(areas_2[10:12])
print(areas_2)

['hallway', 11.25, 'kitchen', 18.0, 'chill zone', 20.0, 'bedroom', 10.75, 'bathroom', 10.5, 'poolhouse', 24.5, 'garage', 15.45]
['hallway', 11.25, 'kitchen', 18.0, 'chill zone', 20.0, 'bedroom', 10.75, 'bathroom', 10.5, 'garage', 15.45]


### Exercise 5-4: Inner workings of lists
In this exercise you'll get some hands-on experience with this. The Python code in the script already creates a list with the name areas and a copy named areas_copy. Next, the first element in the areas_copy list is changed and the areas list is printed out. If you hit Run Code you'll see that, although you've changed areas_copy, the change also takes effect in the areas list. That's because areas and areas_copy point to the same list. If you want to prevent changes in areas_copy from also taking effect in areas, you'll have to do a more explicit copy of the areas list. You can do this with list() or by using [:].

Instructions
1. Change the second command, that creates the variable areas_copy, such that areas_copy is an explicit copy of areas.
2. After your edit, the changes made to areas_copy shouldn't affect areas.

In [ ]:
# Create list areas
areas = [11.25, 18.0, 20.0, 10.75, 9.50]

# Create areas_copy
areas_copy = areas[:]

# Change areas_copy
areas_copy[0] = 5.0

# Print areas
print(areas)

## Chapter 6: Functions
You have already used functions in this course such as type(), but what is a function? A function is a piece of reusable code aimed at solving a particular task. You can call functions instead of having to write code yourself.

#### Example
Let's say you would like to know the maximum height in the family list. Instead of writing code that goes through each value in the list and determines which is the maximum, you can use one of Phytin's built in functions; max().

In [37]:
fam = [1.73, 1.68, 1.71, 1.89]
print(max(fam))

1.89


Functions work kind of like a black box. It receives an input and produces an output. The result of a function can also be assigned to a new variable as follows:

In [38]:
tallest = max(fam)
print(tallest)

1.89


Another of Python's built in functions is round() which receives two inputs or arguments: the number you want to round and the precision with which to round which is how many digits after the decimal point you want to keep.

#### Example

In [39]:
print(round(1.68, 1))

1.7


In [40]:
#In this case, since the desired precision was not specified, Python will round to the closest integer
print(round(1.68))

2


In [41]:
#To better understand a specific Python function such as round(), you can use help() to open up documentation on the function
help(round)

Help on built-in function round in module builtins:

round(number, ndigits=None)
    Round a number to a given precision in decimal digits.
    
    The return value is an integer if ndigits is omitted or None.  Otherwise
    the return value has the same type as the number.  ndigits may be negative.



As it can be observed in the documentation for round(), this function matches the two inputs to two arguments. The first input will be set to the argument number and the second input will be set to the argument ndigits. Then, round() does its calculations to produce an output with number and ndigits as if they're Python variables in a script.

### How do get to know all Python functions?!

Well this is something that is learned through time, but it's important to keep them in mind becuase whenever you're doing a standard task in Python you can be sure that a Pyhton function already exists for that. Remmember the internet is your friend.

### Exercise 6-1: Familiar functions
Out of the box, Python offers a bunch of built-in functions to make your life as a data scientist easier. You already know two such functions: print() and type(). You've also used the functions str(), int(), bool() and float() to switch between data types. These are built-in functions as well. Calling a function is easy. To get the type of 3.0 and store the output as a new variable, result, you can use the following:

- result = type(3.0)

The general recipe for calling functions and saving the result to a variable is thus:

- output = function_name(input)

Instructions
1. Use print() in combination with type() to print out the type of var1.
2. Use len() to get the length of the list var1. Wrap it in a print() call to directly print it out.
3. Use int() to convert var2 to an integer. Store the output as out2.

In [44]:
# Create variables var1 and var2
var1 = [1, 2, 3, 4]
var2 = True

# Print out type of var1
print(type(var1))

# Print out length of var1
print(len(var1))

# Convert var2 to an integer: out2
out2 = int(var2)
print(out2)

<class 'list'>
4
1


### Exercise 6-2: Help!
Maybe you already know the name of a Python function, but you still have to figure out how to use it. Ironically, you have to ask for information about a function with another function: help(). In IPython specifically, you can also use ? before the function name. To get help on the max() function, for example, you can use one of these calls:

- help(max)
- ?max

Use the Shell on the right to open up the documentation on complex(). Which of the following statements is true?

Possible Answers:
- Option 1: complex() takes exactly two arguments: real and [, imag].
- Option 2: complex() takes two arguments: real and imag. Both these arguments are required.
- Option 3: complex() takes two arguments: real and imag. real is a required argument, imag is an optional argument.
- Option 4: complex() takes two arguments: real and imag. If you don't specify imag, it is set to 1 by Python.   

In [46]:
print(help(complex))

Help on class complex in module builtins:

class complex(object)
 |  complex(real=0, imag=0)
 |  
 |  Create a complex number from a real part and an optional imaginary part.
 |  
 |  This is equivalent to (real + imag*1j) where imag defaults to 0.
 |  
 |  Methods defined here:
 |  
 |  __abs__(self, /)
 |      abs(self)
 |  
 |  __add__(self, value, /)
 |      Return self+value.
 |  
 |  __bool__(self, /)
 |      self != 0
 |  
 |  __divmod__(self, value, /)
 |      Return divmod(self, value).
 |  
 |  __eq__(self, value, /)
 |      Return self==value.
 |  
 |  __float__(self, /)
 |      float(self)
 |  
 |  __floordiv__(self, value, /)
 |      Return self//value.
 |  
 |  __format__(...)
 |      complex.__format__() -> str
 |      
 |      Convert to a string according to format_spec.
 |  
 |  __ge__(self, value, /)
 |      Return self>=value.
 |  
 |  __getattribute__(self, name, /)
 |      Return getattr(self, name).
 |  
 |  __getnewargs__(...)
 |  
 |  __gt__(self, value, /)
 | 

Answer: Option 3

### Exercise 6-3: Multiple arguments
In the previous exercise, the square brackets around imag in the documentation showed us that the imag argument is optional. But Python also uses a different way to tell users about arguments being optional. Have a look at the documentation of sorted() by typing help(sorted) in the IPython Shell. You'll see that sorted() takes three arguments: iterable, key and reverse. Key equals none which means that if you don't specify the key argument, it will be None. reverse=False means that if you don't specify the reverse argument, it will be False. In this exercise, you'll only have to specify iterable and reverse, not key. The first input you pass to sorted() will be matched to the iterable argument, but what about the second input? To tell Python you want to specify reverse without changing anything about key, you can use =:

- sorted(___, reverse = ___)

Two lists have been created for you on the right. Can you paste them together and sort them in descending order?
Note: For now, we can understand an iterable as being any collection of objects, e.g. a List.

Instructions
1. Use + to merge the contents of first and second into a new list: full.
2. Call sorted() on full and specify the reverse argument to be True. Save the sorted list as full_sorted.
3. Finish off by printing out full_sorted.

In [47]:
print(help(sorted))

Help on built-in function sorted in module builtins:

sorted(iterable, /, *, key=None, reverse=False)
    Return a new list containing all items from the iterable in ascending order.
    
    A custom key function can be supplied to customize the sort order, and the
    reverse flag can be set to request the result in descending order.

None


In [48]:
# Create lists first and second
first = [11.25, 18.0, 20.0]
second = [10.75, 9.50]

# Paste together first and second: full
full = first[:] + second[:]

# Sort full in descending order: full_sorted
full_sorted = sorted(full, reverse=True)

# Print out full_sorted
print(full_sorted)

[20.0, 18.0, 11.25, 10.75, 9.5]


## Chapter 7: Methods
Built-in functions are only one part of the Python story. We already know about functions to obtain the maximum of a list and the lenght of a list, but what if I'd like to get the index of a specific element or what if I'd like to get a reversing of a list?

In previous sections variables such as sister, height, and fam were created which were of type str, float, and list respectively. These variables types are all recognized as different object by Python which is an object-oriented language (everything is an object). Python object come with a bunch of so-called methods. The term methods is used to refer to all the functions that belong to an object. Methods are called with the dot (.) notation after writing the object from which the method is being called.

#### Example of methods
- str: capitalize(), replace()
- float: bit_length(), conjugate()
- list: index(), count()

#### Example: list methods

In [55]:
print(family_heights)

#Getting the index of the string "mom" in the fam list
print(family_heights.index("mom"))

#Getting the count on the number of times that 1.73 occurs in the fam list
print(family_heights.count(1.73))

['liz', 1.73, 'emma', 1.68, 'mom', 1.71, 'dad', 1.89]
4
1


#### Example: str methods

In [57]:
sister = "liz"

#Capitalizng the first letter of a string
print(sister.capitalize())

#Replacing some parts of the string with other parts
print(sister.replace("z", "sa"))

Liz
lisa


It's important to note that different object can have methods with the same name, but these methods don't necessarly behave similarly. Also, some methods can change the objects they're called on.

#### Example

In [6]:
print(family_heights)

#Appending new data to a list
family_heights.append("me"); family_heights.append(1.62)
print(family_heights)

['liz', 1.73, 'emma', 1.68, 'mom', 1.71, 'dad', 1.89]
['liz', 1.73, 'emma', 1.68, 'mom', 1.71, 'dad', 1.89, 'me', 1.62]


This is pretty useful becuase it allows us to write very concise code to update your data structures on the fly, but it can also be pretty dangerous. Some method calls don't change the objects they're called on, while others do, so watch out.

### Exercise 7-1: String Methods
Strings come with a bunch of methods. Follow the instructions closely to discover some of them. If you want to discover them in more detail, you can always type help(str) in the IPython Shell. A string place has already been created for you to experiment with.

Instructions
1. Use the upper() method on place and store the result in place_up. Use the syntax for calling methods that you learned in the previous section.
2. Print out place and place_up. Did both change?
3. Print out the number of o's on the variable place by calling count() on place and passing the letter 'o' as an input to the method. We're talking about the variable place, not the word "place"!

In [7]:
# string to experiment with: place
place = "poolhouse"

# Use upper() on place: place_up
place_up = place.upper()

# Print out place and place_up
print(place)
print(place_up)

# Print out the number of o's in place
print(place.count("o"))

poolhouse
POOLHOUSE
3


### Exercise 7-2: List Methods (1)
Strings are not the only Python types that have methods associated with them. Lists, floats, integers and booleans are also types that come packaged with a bunch of useful methods. In this exercise, you'll be experimenting with:

- index(): to get the index of the first element of a list that matches its input
- count(): to get the number of times an element appears in a list.

You'll be working on the list with the area of different parts of a house: areas.

Instructions
1. Use the index() method to get the index of the element in areas that is equal to 20.0. Print out this index.
2. Call count() on areas to find out how many times 9.50 appears in the list. Again, simply print out this number.

In [8]:
# Create list areas
areas = [11.25, 18.0, 20.0, 10.75, 9.50]

# Print out the index of the element 20.0
print(areas.index(20.0))

# Print out how often 9.50 appears in areas
print(areas.count(9.50))


2
1


### Exercise 7-3: List Methods (2)
Most list methods will change the list they're called on. Examples are:

- append(): that adds an element to the list it is called on,
- remove(): that removes the first element of a list that matches the input, and
- reverse(): that reverses the order of the elements in the list it is called on.

You'll be working on the list with the area of different parts of the house: areas.

Instructions
1. Use append() twice to add the size of the poolhouse and the garage again: 24.5 and 15.45, respectively. Make sure to add them in this order.
2. Print out areas
3. Use the reverse() method to reverse the order of the elements in areas.
4. Print out areas once more.


In [9]:
# Create list areas
areas = [11.25, 18.0, 20.0, 10.75, 9.50]

# Use append twice to add poolhouse and garage size
areas.append(24.5); areas.append(15.45)

# Print out areas
print(areas)

# Reverse the orders of the elements in areas
areas.reverse()

# Print out areas
print(areas)

[11.25, 18.0, 20.0, 10.75, 9.5, 24.5, 15.45]
[15.45, 24.5, 9.5, 10.75, 20.0, 18.0, 11.25]


## Chapter 8: Packages
By now the power of Python functions and methods has been observed as they allow us to use code snippets from different programs to solve new problems. However, adding all the Python functions and methods available in the same Python distribution would be a mess. There would be lots and lots of code that you would never use. Also, code maintenance would be really problematic. This is were packages come into play. You can think of packages as a directory of Python scripts:

- pkg/
    mod1.py
    mod2.py
    ...

Each script is refered to as a module which specify functions, methods, and new Python types aimed at solving particular problems. There are thousands of Python packages available from the internet. Among them are packages for data science such as numpy (used to efficiently work with arrays), matplotlib (used for data visualization), and scikit-learn (used for machine learning).

Not all of these packages are available in Python by default. To use Python packages, you'll first have to install them on your own system and then, put code in your script to tell Python that you want to use these packages. To execute the installing, you'll have to use pip, a package maintenance system for Python. If you go to the following URL: 

http://pip.readthedocs.org/en/stable/installing/

you can download the file get-pip.py. Next, you go to the terminal and execute "python3 get-pip.py" and you'll be able to use pip and install the Python packages of your choosing. Suppose you want to install the numpy package which will be discussed in the next section, you'll type "python3 pip3 install numpy" and you'll be able to use the numpy package in your Python scripts. However, before you use an install Python package, you should import the package, or the specific module of the package you're going to be using. This is done through the import statement. To import the entire numpy package, you can write "import numpy" on your code. Then, to refer to a function from a package you will use the dot (.) notation after writing the name of the package.

#### Example
Let's say that you wish to use the array function from the numpy package. Then, the following is written: numpy.array([1, 2, 3])

Using the package name and (.) notation as a prefix for each function that you call from a package can be pretty tiring. However, you can also import the package and refer to it with a different name by extending the import statement as follows: import numpy as np. Now, instead of "numpy.array()" you'll have to use np.array to use numpy's array function.

There are also cases were you only need one specific function of the package. Python allows you to make this explicit in your code. Suppose we only wanted to use the array function from the numpy package. Then, the import statement can be changed from "import numpy" to "from numpy import array". This "from import" version to use specific parts of a package can be useful to limit the amount of coding, but you're also loosing some of the context. You loose context because when you import a specific function, you won't have to use the dot (.) notation specifying the package in which the function is found for a function call. As a result, a person reviewing the code might forget that the array function being used is a specific numpy function.

### Exercise 8-1: Import package
As a data scientist, some notions of geometry never hurt. Let's refresh some of the basics. For a fancy clustering algorithm, you want to find the circumference, C, and area, A, of a circle. When the radius of the circle is r, you can calculate C and A as:

- C=2πr
- A=πr^2

To use the constant pi, you'll need the math package. A variable r is already coded in the script. Fill in the code to calculate C and A and see how the print() functions create some nice printouts.

Instructions
1. Import the math package. Now you can access the constant pi with math.pi.
2. Calculate the circumference of the circle and store it in C.
3. Calculate the area of the circle and store it in A.

In [11]:
# Definition of radius
r = 0.43

# Import the math package
import math

# Calculate C
C = 2*math.pi*r

# Calculate A
A = math.pi*r**2

# Build printout
print("Circumference: " + str(C))
print("Area: " + str(A))

Circumference: 2.701769682087222
Area: 0.5808804816487527


### Exercise 8-2: Selective import
General imports, like import math, make all functionality from the math package available to you. However, if you decide to only use a specific part of a package, you can always make your import more selective:

- from math import pi

Let's say the Moon's orbit around planet Earth is a perfect circle, with a radius r (in km) that is defined in the script.

Instructions
1. Perform a selective import from the math package where you only import the radians function.
2. Calculate the distance travelled by the Moon over 12 degrees of its orbit. 3. Assign the result to dist. You can calculate this as r * phi, where r is the radius and phi is the angle in radians. To convert an angle in degrees to an angle in radians, use the radians() function, which you just imported.
4. Print out dist.

In [12]:
# Definition of radius
r = 192500

# Import radians function of math package
from math import radians

# Travel distance of Moon over 12 degrees. Store in dist.
phi = radians(12)
dist = r*phi

# Print out dist
print(dist)

40317.10572106901


### Exercise 8-3: Different ways of importing
There are several ways to import packages and modules into Python. Depending on the import call, you'll have to use different Python code.

Suppose you want to use the function inv(), which is in the linalg subpackage of the scipy package. You want to be able to use this function as follows:

- my_inv([[1,2], [3,4]])

Which import statement will you need in order to run the above code without an error?

Possible Answers:
- Option 1: import scipy
- Option 2: import scipy.linalg
- Option 3: from scipy.linalg import my_inv
- Option 4: from scipy.linalg import inv as my_inv

Answer: Option 4

## Chapter 9: Numpy
By now the power of Python lists has been observed. A list can hold a collection of values with different types at the same time. You can also change, add, and remove elements. However, one feature is missing that is very important for aspiring data scientists. When, analyzing data, you'll often want to carry out fast operations over entire collections of values which lists do not allow.

#### Example: Operations on two lists

In [3]:
height = [1.73, 1.68, 1.71, 1.89, 1.79]
weight = [65.4, 59.2, 63.6, 88.4, 68.7]

If you want to calculate the BMI for each family member you would have to do the following calculation element-wise: weight/height ** 2. However, this operation would not work when using lists. You could do this by going through each list element one after the other and calculating the BMI for each person, but this is not efficient. A more elegant solution is to use numpy or numeric Python. Numpy is a Python package that, among other things, provides an alternative to the regular python list; the numpy array. Numpy array is very similar to the list, but has one additional feature which is that it allows for calculations over entire arrays.

#### Example: Importing and implementing numpy to calculate BMI

In [4]:
import numpy as np

np_height = np.array(height)
np_weight = np.array(weight)

#Executing element-wise calculation of BMI
bmi = np_weight/np_height**2

print(bmi)
print(type(bmi))

[21.85171573 20.97505669 21.75028214 24.7473475  21.44127836]
<class 'numpy.ndarray'>


### Numpy: Remarks
Numpy can do element-wise calculations with its arrays because it assumes that the array contains values of a single type such as floats, booleans, etc. If you try to create an array with multiple types, the resulting numpy array will convert everything to a single type. Also, you should know that the numpy array is simply another Python type which comes with its own unique methods.

#### Example: How numpy arrays differs in behavior to lists

In [5]:
python_list = [1,2,3]
numpy_array = np.array([1,2,3])

#Addition of two lists
print(python_list + python_list)

#Addition of two arrays
print(numpy_array + numpy_array)

[1, 2, 3, 1, 2, 3]
[2 4 6]


### Numpy Subsetting
From another point of view, you can work with numpy arrays pretty much the same as you can would with python lists. You can refer to an element in the array through indexing as with lists as follows:

In [6]:
print(bmi[1])

20.97505668934241


But, with numpy you can also do list subsetting by using boolean logic as follows:

In [7]:
#Getting a boolean array specifying which indexed bmi values are over 23
print(bmi > 23)

[False False False  True False]


Then, the same command can be written in square brackets to get only the bmi values that are over 23 as follows:

In [8]:
#Getting bmi values that are over 23
print(bmi[bmi > 23])

[24.7473475]


### Exercise 9-1: Your First NumPy Array
In this chapter, we're going to dive into the world of baseball. Along the way, you'll get comfortable with the basics of numpy, a powerful package to do data science. A list baseball has already been defined in the Python script, representing the height of some baseball players in centimeters. Can you add some code here and there to create a numpy array from it?

Instructions
1. Import the numpy package as np, so that you can refer to numpy with np.
2. Use np.array() to create a numpy array from baseball. Name this array np_baseball.
3. Print out the type of np_baseball to check that you got it right.

In [9]:
# Create list baseball
baseball = [180, 215, 210, 210, 188, 176, 209, 200]

# Import the numpy package as np
import numpy as np

# Create a numpy array from baseball: np_baseball
np_baseball = np.array(baseball)

# Print out type of np_baseball
print(np_baseball)
print(type(np_baseball))

[180 215 210 210 188 176 209 200]
<class 'numpy.ndarray'>


### Exercise 9-2: Baseball players' height
You are a huge baseball fan. You decide to call the MLB (Major League Baseball) and ask around for some more statistics on the height of the main players. They pass along data on more than a thousand players, which is stored as a regular Python list: height_in. The height is expressed in inches. Can you make a numpy array out of it and convert the units to meters?

height_in is already available and the numpy package is loaded, so you can start straight away (Source: http://wiki.stat.ucla.edu/socr/index.php/SOCR_Data_MLB_HeightsWeights).

Instructions
1. Create a numpy array from height_in. Name this new array np_height_in.
2. Print np_height_in.
3. Multiply np_height_in with 0.0254 to convert all height measurements from inches to meters. Store the new values in a new array, np_height_m.
4. Print out np_height_m and check if the output makes sense.

In [14]:
# height is available as a regular list

# Import pandas
import pandas as pd
mlb = pd.read_csv("http://s3.amazonaws.com/assets.datacamp.com/course/intro_to_python/baseball.csv")
height_in = mlb['Height'].tolist()

# Import numpy
import numpy as np

# Create a numpy array from height_in: np_height_in
np_height_in = np.array(height_in)

# Print out np_height_in
print(np_height_in)

# Convert np_height_in to m: np_height_m
np_height_m = np_height_in*0.0254

# Print np_height_m
print(np_height_m)

[74 74 72 ... 75 75 73]
[1.8796 1.8796 1.8288 ... 1.905  1.905  1.8542]


### Exercise 9-3: Baseball player's BMI
The MLB also offers to let you analyze their weight data. Again, both are available as regular Python lists: height_in and weight_lb. height_in is in inches and weight_lb is in pounds. It's now possible to calculate the BMI of each baseball player. Python code to convert height_in to a numpy array with the correct units is already available in the workspace. Follow the instructions step by step and finish the game!

Instructions
1. Create a numpy array from the weight_lb list with the correct units. Multiply by 0.453592 to go from pounds to kilograms. Store the resulting numpy array as np_weight_kg.
2. Use np_height_m and np_weight_kg to calculate the BMI of each player. Use the following equation: BMI=weight(kg)/height(m)^2
3. Save the resulting numpy array as bmi.
4. Print out bmi.

In [16]:
# height and weight are available as regular lists

# Import pandas
import pandas as pd
mlb = pd.read_csv("http://s3.amazonaws.com/assets.datacamp.com/course/intro_to_python/baseball.csv")
height_in = mlb['Height'].tolist()
weight_lb = mlb['Weight'].tolist()

# Import numpy
import numpy as np

# Create array from height_in with metric units: np_height_m
np_height_m = np.array(height_in) * 0.0254

# Create array from weight_lb with metric units: np_weight_kg
np_weight_kg = np.array(weight_lb) * 0.453592

# Calculate the BMI: bmi
bmi = np_weight_kg/np_height_m**2

# Print out bmi
print(bmi)

[23.11037639 27.60406069 28.48080465 ... 25.62295933 23.74810865
 25.72686361]


### Exercise 9-4: Lightweight baseball players
To subset both regular Python lists and numpy arrays, you can use square brackets:

- x = [4 , 9 , 6, 3, 1]
- x[1]
- import numpy as np
- y = np.array(x)
- y[1]

For numpy specifically, you can also use boolean numpy arrays:

- high = y > 5
- y[high]

The code that calculates the BMI of all baseball players is already included. Follow the instructions and reveal interesting things from the data!

Instructions
1. Create a boolean numpy array: the element of the array should be True if the corresponding baseball player's BMI is below 21. You can use the < operator for this. Name the array light.
2. Print the array light.
3. Print out a numpy array with the BMIs of all baseball players whose BMI is below 21. Use light inside square brackets to do a selection on the bmi array.

In [17]:
# height and weight are available as a regular lists

# Import pandas
import pandas as pd
mlb = pd.read_csv("http://s3.amazonaws.com/assets.datacamp.com/course/intro_to_python/baseball.csv")
height_in = mlb['Height'].tolist()
weight_lb = mlb['Weight'].tolist()

# Import numpy
import numpy as np

# Calculate the BMI: bmi
np_height_m = np.array(height_in) * 0.0254
np_weight_kg = np.array(weight_lb) * 0.453592
bmi = np_weight_kg / np_height_m ** 2

# Create the light array
light = bmi < 21

# Print out light
print(light)

# Print out BMIs of all baseball players whose BMI is below 21
print(bmi[light])

[False False False ... False False False]
[20.54255679 20.54255679 20.69282047 20.69282047 20.34343189 20.34343189
 20.69282047 20.15883472 19.4984471  20.69282047 20.9205219 ]


### Exercise 9-5: NumPy Side Effects
Numpy is great for doing vector arithmetic. If you compare its functionality with regular Python lists, however, some things have changed.

First of all, numpy arrays cannot contain elements with different types. If you try to build such a list, some of the elements' types are changed to end up with a homogeneous list. This is known as type coercion.

Second, the typical arithmetic operators, such as +, -, * and / have a different meaning for regular Python lists and numpy arrays. Have a look at this line of code:
- np.array([True, 1, 2]) + np.array([3, 4, False])

Can you tell which code chunk builds the exact same Python object? The numpy package is already imported as np, so you can start experimenting in the IPython Shell straight away!

Possible Answers
- Option 1: np.array([True, 1, 2, 3, 4, False])
- Option 2: np.array([4, 3, 0]) + np.array([0, 2, 2])
- Option 3: np.array([1, 1, 2]) + np.array([3, 4, -1])
- Option 4: np.array([0, 1, 2, 3, 4, 5])

Answer: Option 2

### Exercise 9-6: Subsetting NumPy Arrays
You've seen it with your own eyes: Python lists and numpy arrays sometimes behave differently. Luckily, there are still certainties in this world. For example, subsetting (using the square bracket notation on lists or arrays) works exactly the same. To see this for yourself, try the following lines of code in the IPython Shell:

- x = ["a", "b", "c"]
- x[1]
- np_x = np.array(x)
- np_x[1]

The script on the right already contains code that imports numpy as np, and stores both the height and weight of the MLB players as numpy arrays.

Instructions
1. Subset np_weight_lb by printing out the element at index 50.
2. Print out a sub-array of np_height_in that contains the elements at index 100 up to and including index 110.


In [19]:
# height and weight are available as a regular lists

# Import pandas
import pandas as pd
mlb = pd.read_csv("http://s3.amazonaws.com/assets.datacamp.com/course/intro_to_python/baseball.csv")
height_in = mlb['Height'].tolist()
weight_lb = mlb['Weight'].tolist()

# Import numpy
import numpy as np

# Store weight and height lists as numpy arrays
np_weight_lb = np.array(weight_lb)
np_height_in = np.array(height_in)

# Print out the weight at index 50
print(np_weight_lb[50])

# Print out sub-array of np_height_in: index 100 up to and including index 110
print(np_height_in[100:111])

200
[73 74 72 73 69 72 73 75 75 73 72]


## Chapter 10: 2D Numpy Arrays

Let's recolect the numpy arrays from the previous sections. If you ask Python for the type of these arrays it will say they are of type numpy.ndarray. The numpy. prefix tells you that it's a type that was defined in the numpy package. Then, ndarray stands for n-dimensional array. The arrays in np_height and np_weight are 1-dimensional, but it's possible to create arrays with more than one dimension. In this section, we'll work with 2-dimensional numpy arrays.

You can create a numpy 2D array from a regular Python list of lists as follows:

In [3]:
import numpy as np

np_2d = np.array([[1.73, 1.68, 1.71, 1.89, 1.79], [65.4, 59.2, 63.6, 88.4, 68.7]])
print(np_2d)

[[ 1.73  1.68  1.71  1.89  1.79]
 [65.4  59.2  63.6  88.4  68.7 ]]


Then, from the following command we can see that np_2d has 2 rows and 5 columns:

In [6]:
print(np_2d.shape)

(2, 5)


Shape is an attribute of the np_2d array that can give you more information of what the data structure looks like. Note that the syntax for calling an attribute is similar to the one that is used for calling a method, but these are not the same.

### Subsetting
You can think of the numpy 2D array as an improved list of lists since you can preform calculations on the array and implement more advanced subsetting methods.

#### Example
Suppose you want the first row and then the third element in the row. In that case, the following command is written:

In [9]:
print(np_2d[0][2])

#Different command, same output
print(np_2d[0,2])

1.71
1.71


#### Example
Suppose you want to select the height and weight of the second and thrid family member. In that case, the following command is written:

In [15]:
print("Height and weight of 2nd family member: ", str(np_2d[0][1]), str(np_2d[1][1]))
print("Height and weight of 3nd family member: ", str(np_2d[0][2]), str(np_2d[1][2]))

#Different command, same output.
#Note that : before the comma is used because I want to output both rows
print(np_2d[:, 1:3])

Height and weight of 2nd family member:  1.68 59.2
Height and weight of 3nd family member:  1.71 63.6
[[ 1.68  1.71]
 [59.2  63.6 ]]


Similarly you can select the weight of all family members as follows:

In [16]:
print(np_2d[1,:])

[65.4 59.2 63.6 88.4 68.7]


### Exercise 10-1: Your First 2D NumPy Array
Before working on the actual MLB data, let's try to create a 2D numpy array from a small list of lists. In this exercise, baseball is a list of lists. The main list contains 4 elements. Each of these elements is a list containing the height and the weight of 4 baseball players, in this order. baseball is already coded for you in the script.

Instructions
1. Use np.array() to create a 2D numpy array from baseball. Name it np_baseball.
2. Print out the type of np_baseball.
3. Print out the shape attribute of np_baseball. Use np_baseball.shape.

In [18]:
# Create baseball, a list of lists
baseball = [[180, 78.4],
            [215, 102.7],
            [210, 98.5],
            [188, 75.2]]

# Import numpy
import numpy as np

# Create a 2D numpy array from baseball: np_baseball
np_baseball = np.array(baseball)

# Print out the type of np_baseball
print(type(np_baseball))
print(np_baseball.shape)

<class 'numpy.ndarray'>
(4, 2)


### Exercise 10-2: Baseball data in 2D form
You have another look at the MLB data and realize that it makes more sense to restructure all this information in a 2D numpy array. This array should have 1015 rows, corresponding to the 1015 baseball players you have information on, and 2 columns (for height and weight).

The MLB was, again, very helpful and passed you the data in a different structure, a Python list of lists. In this list of lists, each sublist represents the height and weight of a single baseball player. The name of this embedded list is baseball.

Can you store the data as a 2D array to unlock numpy's extra functionality?

Instructions
1. Use np.array() to create a 2D numpy array from baseball. Name it np_baseball.
2. Print out the shape attribute of np_baseball.

In [20]:
# baseball is available as a regular list of lists

# Import pandas
import pandas as pd
baseball = pd.read_csv("http://s3.amazonaws.com/assets.datacamp.com/course/intro_to_python/baseball.csv")

# Import numpy package
import numpy as np

# Create a 2D numpy array from baseball: np_baseball
np_baseball = np.array(baseball)

# Print out the shape of np_baseball
print(np_baseball.shape)

(1015, 7)


### Exercise 10-3: Subsetting 2D NumPy Arrays
If your 2D numpy array has a regular structure, i.e. each row and column has a fixed number of values, complicated ways of subsetting become very easy. Have a look at the code below where the elements "a" and "c" are extracted from a list of lists.

- #regular list of lists
- x = [["a", "b"], ["c", "d"]]
- [x[0][0], x[1][0]]

- #numpy
- import numpy as np
- np_x = np.array(x)
- np_x[:,0]

For regular Python lists, this is a real pain. For 2D numpy arrays, however, it's pretty intuitive! The indexes before the comma refer to the rows, while those after the comma refer to the columns. The : is for slicing; in this example, it tells Python to include all rows.

The code that converts the pre-loaded baseball list to a 2D numpy array is already in the script. The first column contains the players' height in inches and the second column holds player weight, in pounds. Add some lines to make the correct selections. Remember that in Python, the first element is at index 0!

Instructions
1. Print out the 50th row of np_baseball.
2. Make a new variable, np_weight_lb, containing the entire second column of np_baseball.
3. Select the height (first column) of the 124th baseball player in np_baseball and print it out.

In [24]:
# baseball is available as a regular list of lists

# Import pandas package
import pandas as pd
baseball = pd.read_csv("http://s3.amazonaws.com/assets.datacamp.com/course/intro_to_python/baseball.csv")[['Height', 'Weight']].values.tolist()

# Import numpy package
import numpy as np

# Create np_baseball (2 cols)
np_baseball = np.array(baseball)

# Print out the 50th row of np_baseball
print(np_baseball[49, :])

# Select the entire second column of np_baseball: np_weight_lb
np_weight_lb = np_baseball[:, 1]

# Print out height of 124th player among other things
print(np_baseball[123, 0])

[ 70 195]
75


### Exercise 10-4: 2D Arithmetic
Remember how you calculated the Body Mass Index for all baseball players? numpy was able to perform all calculations element-wise (i.e. element by element). For 2D numpy arrays this isn't any different! You can combine matrices with single numbers, with vectors, and with other matrices. Execute the code below in the IPython shell and see if you understand:

- import numpy as np
- np_mat = np.array([[1, 2],[3, 4],[5, 6]])
- np_mat * 2
- np_mat + np.array([10, 10])
- np_mat + np_mat
- np_baseball is coded for you; it's again a 2D numpy array with 3 columns representing height (in inches), weight (in pounds) and age (in years).

Instructions
1. You managed to get hold of the changes in height, weight and age of all baseball players. It is available as a 2D numpy array, updated. Add np_baseball and updated and print out the result.
2. You want to convert the units of height and weight to metric (meters and kilograms respectively). As a first step, create a numpy array with three values: 0.0254, 0.453592 and 1. Name this array conversion.
3. Multiply np_baseball with conversion and print out the result.

In [27]:
# baseball is available as a regular list of lists
# updated is available as 2D numpy array

# Import packages
baseball = pd.read_csv("http://s3.amazonaws.com/assets.datacamp.com/course/intro_to_python/baseball.csv")[['Height', 'Weight', 'Age']].values.tolist()
updated = pd.read_csv("http://s3.amazonaws.com/assets.datacamp.com/course/intro_to_python/update.csv", header = None)

# Create np_baseball (3 cols)
np_baseball = np.array(baseball)

# Print out addition of np_baseball and updated
print(np_baseball + updated)

# Create numpy array: conversion
conversion = np.array([[0.0254,0.453592,1]])

# Print out product of np_baseball and conversion
print(np_baseball*conversion)

              0           1      2
0     75.230356  168.837751  23.99
1     75.026143  231.097323  35.69
2     73.154423  215.081676  31.78
3     72.644275  204.904619  36.43
4     74.005901  190.243427  36.71
...         ...         ...    ...
1010  73.987187  195.666656  26.08
1011  75.205969  175.472193  26.73
1012  76.093499  209.238908  26.19
1013  75.822857  172.218000  32.01
1014  73.994842  203.144027  28.92

[1015 rows x 3 columns]
[[ 1.8796  81.64656 22.99   ]
 [ 1.8796  97.52228 34.69   ]
 [ 1.8288  95.25432 30.78   ]
 ...
 [ 1.905   92.98636 25.19   ]
 [ 1.905   86.18248 31.01   ]
 [ 1.8542  88.45044 27.92   ]]


## Chapter 11: Numpy - Basic Statistics
A typical first step in analysing your data is getting to know your data in the first place. For the numpy arrays from previous sections this is a simple task because they didn't contain much data. However, as a data scientist you'll be working with thousands, if not millions or billions of numbers.

Imagine you conduct a city-wide survey where you ask 5000 adults about their height and weight. In that case, you'll probably end up with a 2D numpy array called np_city with 5000 rows corresponding to the 5000 people and two columns corresponding to height and weight respectively. However, the individual numbers won't give you any insights into the data. True insights on the data can only be obtained through summary statistics about the data. The numpy package is also useful in this case providing functions such as:

- np.mean(np_city[:, 0]): Gets average height of 5000 people.
- np.median(np_city[:, 0): Gets height of middle person if you sort heights from small to tall.
- np.corrcoef(np_city[:,0], np_city[:,1]): Gets Pearson product-moment correlation coefficients in matrix form which, in this case, specifies the correlation between height and weight. The correlation coefficient is a measure of the direction and strength of a linear relationship among variables. Correlation coefficients range from -1 to +1 where: -1 means perfect negative relationship, 0 means no relationship, and +1 means perfect positive relationship.
- np.std(np_city[:,0]: Gets the standard deviation of height

Numpy also features more basic functions such as sum() and sort() which also exist in the basic Python distribution. However, the big difference here is speed since numpy enforces a single data type in an array.

Furthermore, numpy allows for the generation of data through the np.random.normal() function which has three arguments: distribution mean, distribution standard deviation, and number of samples.

#### Example: Generating data for np_city

In [28]:
#sampling two random distributions 5000 times to create height array
height = np.round(np.random.normal(1.75, 0.20, 5000), 2)

#sampling two random distributions 5000 times to create weight array
weight = np.round(np.random.normal(60.32, 15, 5000), 2)
np_city = np.column_stack((height, weight))
print(np_city)

[[ 1.67 64.66]
 [ 1.7  54.81]
 [ 1.24 74.53]
 ...
 [ 1.64 66.77]
 [ 1.62 58.6 ]
 [ 1.6  51.39]]


### Exercise 11-1: Average versus median
You now know how to use numpy functions to get a better feeling for your data. It basically comes down to importing numpy and then calling several simple functions on the numpy arrays:

- import numpy as np
- x = [1, 4, 8, 10, 12]
- np.mean(x)
- np.median(x)

The baseball data is available as a 2D numpy array with 3 columns (height, weight, age) and 1015 rows. The name of this numpy array is np_baseball. After restructuring the data, however, you notice that some height values are abnormally high. Follow the instructions and discover which summary statistic is best suited if you're dealing with so-called outliers.

Instructions
1. Create numpy array np_height_in that is equal to first column of np_baseball.
2. Print out the mean of np_height_in.
3. Print out the median of np_height_in.

In [32]:
# np_baseball is available

import pandas as pd
np_baseball = pd.read_csv("http://s3.amazonaws.com/assets.datacamp.com/course/intro_to_python/baseball.csv")[['Height', 'Weight', 'Age']].values
np_baseball[slice(0, 1015, 50), 0] = np_baseball[slice(0, 1015, 50), 0]*1000

# Import numpy
import numpy as np

# Create np_height_in from np_baseball
np_height_in = np_baseball[:, 0]

# Print out the mean of np_height_in
print(np.mean(np_height_in))

# Print out the median of np_height_in
print(np.median(np_height_in))

1586.4610837438424
74.0


An average height of 1586 inches, that doesn't sound right, does it? However, the median does not seem affected by the outliers: 74 inches makes perfect sense. It's always a good idea to check both the median and the mean, to get an idea about the overall distribution of the entire dataset.

### Exercise 11-2: Explore the baseball data
Because the mean and median are so far apart, you decide to complain to the MLB. They find the error and send the corrected data over to you. It's again available as a 2D Numpy array np_baseball, with three columns. The Python script on the right already includes code to print out informative messages with the different summary statistics. Can you finish the job?

Instructions
1. The code to print out the mean height is already included. Complete the code for the median height. Replace None with the correct code.
2. Use np.std() on the first column of np_baseball to calculate stddev. Replace None with the correct code.
3. Do big players tend to be heavier? Use np.corrcoef() to store the correlation between the first and second column of np_baseball in corr. Replace None with the correct code.

In [34]:
# np_baseball is available

import pandas as pd
np_baseball = pd.read_csv("http://s3.amazonaws.com/assets.datacamp.com/course/intro_to_python/baseball.csv")[['Height', 'Weight', 'Age']].values

# Import numpy
import numpy as np

# Print mean height (first column)
avg = np.mean(np_baseball[:,0])
print("Average: " + str(avg))

# Print median height. Replace 'None'
med = np.median(np_baseball[:,0])
print("Median: " + str(med))

# Print out the standard deviation on height. Replace 'None'
stddev = np.std(np_baseball[:,0])
print("Standard Deviation: " + str(stddev))

# Print out correlation between first and second column. Replace 'None'
corr = np.corrcoef(np_baseball[:,0], np_baseball[:,1])
print("Correlation: " + str(corr))

Average: 73.6896551724138
Median: 74.0
Standard Deviation: 2.312791881046546
Correlation: [[1.         0.53153932]
 [0.53153932 1.        ]]


### Exercise 11-3: Blend it all together
In the last few exercises you've learned everything there is to know about heights and weights of baseball players. Now it's time to dive into another sport: soccer. You've contacted FIFA for some data and they handed you two lists. The lists are the following:

- positions = ['GK', 'M', 'A', 'D', ...]
- heights = [191, 184, 185, 180, ...]

Each element in the lists corresponds to a player. The first list, positions, contains strings representing each player's position. The possible positions are: 'GK' (goalkeeper), 'M' (midfield), 'A' (attack) and 'D' (defense). The second list, heights, contains integers representing the height of the player in cm. The first player in the lists is a goalkeeper and is pretty tall (191 cm).

You're fairly confident that the median height of goalkeepers is higher than that of other players on the soccer field. Some of your friends don't believe you, so you are determined to show them using the data you received from FIFA and your newly acquired Python skills.

Instructions
1. Convert heights and positions, which are regular lists, to numpy arrays. Call them np_heights and np_positions.
2. Extract all the heights of the goalkeepers. You can use a little trick here: use np_positions == 'GK' as an index for np_heights. Assign the result to gk_heights.
3. Extract all the heights of all the other players. This time use np_positions != 'GK' as an index for np_heights. Assign the result to other_heights.
4. Print out the median height of the goalkeepers using np.median(). Replace None with the correct code.
5. Do the same for the other players. Print out their median height. Replace None with the correct code.

In [44]:
# heights and positions are available as lists

# Import pandas
import pandas as pd
fifa = pd.read_csv("https://assets.datacamp.com/production/repositories/288/datasets/026a5211b906ac118a09b1a0dbf7df48faafb379/fifa.csv", skipinitialspace=True, usecols=['position', 'height'])
positions = list(fifa.position)
heights = list(fifa.height)

# Import numpy
import numpy as np

# Convert positions and heights to numpy arrays: np_positions, np_heights
np_positions = np.array(positions)
np_heights = np.array(heights)

# Heights of the goalkeepers: gk_heights
gk_heights = np_heights[np_positions == 'GK']

# Heights of the other players: other_heights
other_heights = np_heights[np_positions != 'GK']

# Print out the median height of goalkeepers. Replace 'None'
print("Median height of goalkeepers: " + str(np.median(gk_heights)))

# Print out the median height of other players. Replace 'None'
print("Median height of other players: " + str(np.median(other_heights)))

Median height of goalkeepers: 188.0
Median height of other players: 181.0
